From 526d98097255d125e4aff129e63ec0696482dbbf Mon Sep 17 00:00:00 2001
From: Chen Shunqing <csq@rock-chips.com>
Date: Mon, 22 Jan 2024 07:23:37 +0000
Subject: [PATCH 7/9] media: i2c: rk628: add hdmirxphy debugfs

Type: Function
Redmine ID: N/A
Associated modifications: N/A
Test: cat /d/rk628/x-0050/registers/hdmirxphy
      echo 0x2 0x1680 > /d/rk628/x-0050/registers/hdmirxphy

Change-Id: Id395cb5c1c373be00ae4916d86f5f6b13c631c1f
Signed-off-by: Chen Shunqing <csq@rock-chips.com>
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/media/i2c/rk628/rk628.c        |  1 +
 drivers/media/i2c/rk628/rk628_hdmirx.c | 68 ++++++++++++++++++++++++++
 drivers/media/i2c/rk628/rk628_hdmirx.h |  1 +
 3 files changed, 70 insertions(+)

diff --git a/drivers/media/i2c/rk628/rk628.c b/drivers/media/i2c/rk628/rk628.c
index 3ba5fb3854016a..01a95757d897ed 100644
--- a/drivers/media/i2c/rk628/rk628.c
+++ b/drivers/media/i2c/rk628/rk628.c
@@ -592,6 +592,7 @@ static void rk628_debugfs_register_create(struct rk628 *rk628)
 			continue;
 		debugfs_create_file(reg->name, 0600, dir, rk628, &rk628_reg_fops);
 	}
+	rk628_hdmirx_phy_debugfs_register_create(rk628, dir);
 }
 
 void rk628_debugfs_create(struct rk628 *rk628)
diff --git a/drivers/media/i2c/rk628/rk628_hdmirx.c b/drivers/media/i2c/rk628/rk628_hdmirx.c
index bc1cff242ca2a8..fca47fc262195f 100644
--- a/drivers/media/i2c/rk628/rk628_hdmirx.c
+++ b/drivers/media/i2c/rk628/rk628_hdmirx.c
@@ -1491,6 +1491,74 @@ bool rk628_hdmirx_is_signal_change_ists(struct rk628 *rk628)
 }
 EXPORT_SYMBOL(rk628_hdmirx_is_signal_change_ists);
 
+static int rk628_hdmirx_phy_reg_show(struct seq_file *s, void *v)
+{
+	struct rk628 *rk628 = s->private;
+	unsigned int i;
+
+	seq_printf(s, "rk628_%s:\n", file_dentry(s->file)->d_iname);
+
+	for (i = 0; i <= 0xb7; i++)
+		seq_printf(s, "0x%02x: %08x\n", i, hdmirxphy_read(rk628, i));
+
+	return 0;
+}
+
+static ssize_t rk628_hdmirx_phy_reg_write(struct file *file, const char __user *buf,
+			       size_t count, loff_t *ppos)
+{
+	struct rk628 *rk628 = file->f_path.dentry->d_inode->i_private;
+	u32 addr;
+	u32 val;
+	char kbuf[25];
+	int ret;
+
+	if (count >= sizeof(kbuf))
+		return -ENOSPC;
+
+	if (copy_from_user(kbuf, buf, count))
+		return -EFAULT;
+
+	kbuf[count] = '\0';
+
+	ret = sscanf(kbuf, "%x%x", &addr, &val);
+	if (ret != 2)
+		return -EINVAL;
+
+	if (addr > 0xb7)
+		return -EINVAL;
+
+	hdmirxphy_write(rk628, addr, val);
+
+	return count;
+}
+
+static int rk628_hdmirx_phy_reg_open(struct inode *inode, struct file *file)
+{
+	struct rk628 *rk628 = inode->i_private;
+
+	return single_open(file, rk628_hdmirx_phy_reg_show, rk628);
+}
+
+static const struct file_operations rk628_hdmirx_phy_reg_fops = {
+	.owner          = THIS_MODULE,
+	.open           = rk628_hdmirx_phy_reg_open,
+	.read           = seq_read,
+	.write          = rk628_hdmirx_phy_reg_write,
+	.llseek         = seq_lseek,
+	.release        = single_release,
+};
+void rk628_hdmirx_phy_debugfs_register_create(struct rk628 *rk628, struct dentry *dir)
+{
+	if (rk628->version < RK628F_VERSION)
+		return;
+	if (IS_ERR(dir))
+		return;
+
+	debugfs_create_file("hdmirxphy", 0600, dir, rk628, &rk628_hdmirx_phy_reg_fops);
+}
+EXPORT_SYMBOL(rk628_hdmirx_phy_debugfs_register_create);
+
 MODULE_AUTHOR("Shunqing Chen <csq@rock-chips.com>");
 MODULE_DESCRIPTION("Rockchip RK628 driver");
 MODULE_LICENSE("GPL v2");
diff --git a/drivers/media/i2c/rk628/rk628_hdmirx.h b/drivers/media/i2c/rk628/rk628_hdmirx.h
index 88f9a4409503cd..1c91b1bb27685d 100644
--- a/drivers/media/i2c/rk628/rk628_hdmirx.h
+++ b/drivers/media/i2c/rk628/rk628_hdmirx.h
@@ -523,4 +523,5 @@ void rk628_hdmirx_cec_unregister(struct rk628_hdmirx_cec *cec);
 void rk628_hdmirx_cec_hpd(struct rk628_hdmirx_cec *cec, bool en);
 void rk628_hdmirx_cec_state_reconfiguration(struct rk628 *rk628,
 					    struct rk628_hdmirx_cec *cec);
+void rk628_hdmirx_phy_debugfs_register_create(struct rk628 *rk628, struct dentry *dir);
 #endif
-- 
2.17.1

